File: ios/Photo/Export/PhotoExporter.swift
import ImageIO
import UIKit

struct PhotoExportResult {
  let uri: String
  let width: Int
  let height: Int
  let fileSize: Int
  let mimeType: String
}

struct PhotoExportError: Error {
  let code: String
  let message: String
}

/// Applies a `PhotoTransformState` to the full-resolution source photo and
/// writes the result to a new file under the app's temporary directory. The
/// source file is never modified. WebP export is not implemented on iOS
/// (no first-party encoder); JPEG and PNG are supported.
enum PhotoExporter {
  private static let defaultMaxDimension: CGFloat = 4096

  static func export(
    sourceUri: String,
    transform: PhotoTransformState,
    adjustments: PhotoAdjustments,
    layers: [PhotoLayer],
    exportOptions: [String: Any]?
  ) throws -> PhotoExportResult {
    guard let path = SourceResolver.resolvePath(sourceUri: sourceUri, tempPrefix: "pve_export") else {
      throw PhotoExportError(code: "E_SOURCE_NOT_FOUND", message: "The selected photo could not be found.")
    }
    let url = URL(fileURLWithPath: path)
    guard let source = CGImageSourceCreateWithURL(url as CFURL, nil) else {
      throw PhotoExportError(code: "E_SOURCE_UNREADABLE", message: "The selected photo could not be decoded.")
    }

    let maxWidth = (exportOptions?["maxWidth"] as? NSNumber)?.doubleValue.map { CGFloat($0) }
    let maxHeight = (exportOptions?["maxHeight"] as? NSNumber)?.doubleValue.map { CGFloat($0) }
    let decodeTarget = max(maxWidth ?? defaultMaxDimension, maxHeight ?? defaultMaxDimension) * 2

    let thumbnailOptions: [CFString: Any] = [
      kCGImageSourceCreateThumbnailFromImageAlways: true,
      kCGImageSourceThumbnailMaxPixelSize: decodeTarget,
      kCGImageSourceCreateThumbnailWithTransform: true,
    ]
    guard let cgImage = CGImageSourceCreateThumbnailAtIndex(source, 0, thumbnailOptions as CFDictionary) else {
      throw PhotoExportError(code: "E_OUT_OF_MEMORY", message: "The photo is too large to process on this device.")
    }
    let upright = UIImage(cgImage: cgImage)

    let transformed = PhotoEditSession.applyTransform(upright, state: transform)
    let adjusted = PhotoAdjustmentRenderer.apply(transformed, adjustments: adjustments)
    let layered = PhotoLayerRenderer.render(adjusted, layers: layers) { uri in resolveImageLayer(uri) }
    let cropped = crop(layered, transform: transform)
    let resized = resize(cropped, maxWidth: maxWidth, maxHeight: maxHeight)

    let format = (exportOptions?["imageFormat"] as? String)?.lowercased() ?? "jpeg"
    let quality = qualityFor(exportOptions?["quality"] as? String ?? "high")

    let outputDir = FileManager.default.temporaryDirectory.appendingPathComponent("photovideoeditor", isDirectory: true)
    try? FileManager.default.createDirectory(at: outputDir, withIntermediateDirectories: true)

    let data: Data?
    let mimeType: String
    let ext: String
    switch format {
    case "png":
      data = resized.pngData()
      mimeType = "image/png"
      ext = "png"
    default:
      // WebP encoding is not implemented on iOS; unsupported formats fall back to JPEG.
      data = resized.jpegData(compressionQuality: quality)
      mimeType = "image/jpeg"
      ext = "jpg"
    }
    guard let outputData = data else {
      throw PhotoExportError(code: "E_EXPORT_FAILED", message: "Unable to encode the exported photo.")
    }

    let outputFile = outputDir.appendingPathComponent(UUID().uuidString + "." + ext)
    do {
      try outputData.write(to: outputFile)
    } catch {
      throw PhotoExportError(code: "E_EXPORT_FAILED", message: "Unable to write the exported photo.")
    }

    let attributes = try? FileManager.default.attributesOfItem(atPath: outputFile.path)
    let fileSize = (attributes?[.size] as? Int) ?? outputData.count

    return PhotoExportResult(
      uri: outputFile.absoluteString,
      width: Int(resized.size.width),
      height: Int(resized.size.height),
      fileSize: fileSize,
      mimeType: mimeType
    )
  }

  /// Export-time, one-shot resolver for sticker-uri and overlay-uri layers alike (no cache needed —
  /// export runs once per invocation).
  private static func resolveImageLayer(_ uri: String) -> UIImage? {
    guard let path = SourceResolver.resolvePath(sourceUri: uri, tempPrefix: "pve_layer_export") else { return nil }
    return UIImage(contentsOfFile: path)
  }

  private static func crop(_ image: UIImage, transform: PhotoTransformState) -> UIImage {
    guard transform.cropLeft > 0 || transform.cropTop > 0 || transform.cropRight < 1 || transform.cropBottom < 1 else { return image }
    guard let cgImage = image.cgImage else { return image }
    let width = CGFloat(cgImage.width)
    let height = CGFloat(cgImage.height)
    let fullRect = CGRect(x: 0, y: 0, width: width, height: height)
    let rect = CGRect(
      x: transform.cropLeft * width,
      y: transform.cropTop * height,
      width: (transform.cropRight - transform.cropLeft) * width,
      height: (transform.cropBottom - transform.cropTop) * height
    ).integral.intersection(fullRect)
    guard rect.width > 0, rect.height > 0, let croppedImage = cgImage.cropping(to: rect) else { return image }
    return UIImage(cgImage: croppedImage)
  }

  private static func resize(_ image: UIImage, maxWidth: CGFloat?, maxHeight: CGFloat?) -> UIImage {
    guard maxWidth != nil || maxHeight != nil else { return image }
    let widthLimit = maxWidth ?? image.size.width
    let heightLimit = maxHeight ?? image.size.height
    guard image.size.width > widthLimit || image.size.height > heightLimit else { return image }
    let scale = min(widthLimit / image.size.width, heightLimit / image.size.height)
    let targetSize = CGSize(width: max(1, image.size.width * scale), height: max(1, image.size.height * scale))
    let renderer = PhotoEditSession.pixelRenderer(size: targetSize)
    return renderer.image { _ in image.draw(in: CGRect(origin: .zero, size: targetSize)) }
  }

  private static func qualityFor(_ preset: String) -> CGFloat {
    switch preset {
    case "low": return 0.5
    case "medium": return 0.75
    case "original": return 1.0
    default: return 0.9
    }
  }
}

File: android/src/main/java/com/photovideoeditor/photo/export/PhotoExporter.kt
package com.photovideoeditor.photo.export

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import android.net.Uri
import android.os.Build
import androidx.exifinterface.media.ExifInterface
import com.photovideoeditor.files.SourceResolver
import com.photovideoeditor.photo.render.PhotoAdjustmentRenderer
import com.photovideoeditor.photo.render.PhotoAdjustments
import com.photovideoeditor.photo.render.PhotoLayer
import com.photovideoeditor.photo.render.PhotoLayerRenderer
import com.photovideoeditor.photo.render.PhotoTransformState
import org.json.JSONObject
import java.io.File
import java.io.FileOutputStream
import java.util.UUID
import kotlin.math.roundToInt

data class PhotoExportResult(
  val uri: String,
  val width: Int,
  val height: Int,
  val fileSize: Long,
  val mimeType: String
)

class PhotoExportException(val code: String, message: String) : Exception(message)

/**
 * Applies a [PhotoTransformState] to the full-resolution source photo and
 * writes the result to a new file in the app cache dir. The source file is
 * never modified. Bitmaps are downsampled to the requested export bounds (or
 * a safe default) to avoid decoding an unbounded full-resolution image twice.
 */
object PhotoExporter {
  private const val DEFAULT_MAX_DIMENSION = 4096

  fun export(
    context: Context,
    sourceUri: String,
    transform: PhotoTransformState,
    adjustments: PhotoAdjustments,
    layers: List<PhotoLayer>,
    exportOptions: JSONObject?
  ): PhotoExportResult {
    val path = SourceResolver.resolvePath(context, sourceUri, "pve_export")
      ?: throw PhotoExportException("E_SOURCE_NOT_FOUND", "The selected photo could not be found.")

    val exif = try { ExifInterface(path) } catch (_: Exception) { null }
    val exifOrientation = exif?.getAttributeInt(ExifInterface.TAG_ORIENTATION, ExifInterface.ORIENTATION_NORMAL)
      ?: ExifInterface.ORIENTATION_NORMAL

    val maxWidth = exportOptions?.optInt("maxWidth", 0)?.takeIf { it > 0 }
    val maxHeight = exportOptions?.optInt("maxHeight", 0)?.takeIf { it > 0 }

    val bounds = BitmapFactory.Options().apply { inJustDecodeBounds = true }
    BitmapFactory.decodeFile(path, bounds)
    if (bounds.outWidth <= 0 || bounds.outHeight <= 0) {
      throw PhotoExportException("E_SOURCE_UNREADABLE", "The selected photo could not be decoded.")
    }

    val decodeTarget = maxOf(maxWidth ?: DEFAULT_MAX_DIMENSION, maxHeight ?: DEFAULT_MAX_DIMENSION)
    val sampleSize = computeSampleSize(bounds.outWidth, bounds.outHeight, decodeTarget)
    val decoded = BitmapFactory.decodeFile(path, BitmapFactory.Options().apply { inSampleSize = sampleSize })
      ?: throw PhotoExportException("E_OUT_OF_MEMORY", "The photo is too large to process on this device.")

    val upright = applyExifOrientation(decoded, exifOrientation)
    val transformed = applyTransform(upright, transform)
    val adjusted = PhotoAdjustmentRenderer.apply(transformed, adjustments)
    val layered = PhotoLayerRenderer.render(adjusted, layers) { uri -> resolveImageUri(context, uri) }
    val cropped = cropBitmap(layered, transform)
    val resized = resizeToFit(cropped, maxWidth, maxHeight)

    val format = exportOptions?.optString("imageFormat", "jpeg")?.lowercase() ?: "jpeg"
    val quality = qualityFor(exportOptions?.optString("quality", "high") ?: "high")
    val (compressFormat, mimeType, extension) = encodingFor(format)

    val outputDir = File(context.cacheDir, "photovideoeditor").apply { mkdirs() }
    val outputFile = File(outputDir, "${UUID.randomUUID()}.$extension")
    FileOutputStream(outputFile).use { stream ->
      if (!resized.compress(compressFormat, quality, stream)) {
        throw PhotoExportException("E_EXPORT_FAILED", "Unable to encode the exported photo.")
      }
    }

    if (format == "jpeg" && (exportOptions?.optBoolean("preserveMetadata", true) != false)) {
      preserveMetadata(exif, outputFile.absolutePath)
    }

    if (decoded !== upright) decoded.recycle()
    if (upright !== transformed) upright.recycle()
    if (transformed !== adjusted) transformed.recycle()
    if (adjusted !== layered) adjusted.recycle()
    if (layered !== cropped) layered.recycle()
    if (cropped !== resized) cropped.recycle()

    return PhotoExportResult(
      uri = Uri.fromFile(outputFile).toString(),
      width = resized.width,
      height = resized.height,
      fileSize = outputFile.length(),
      mimeType = mimeType
    )
  }

  private fun resolveImageUri(context: Context, uri: String): Bitmap? {
    val path = SourceResolver.resolvePath(context, uri, "pve_layer_image_export") ?: return null
    return BitmapFactory.decodeFile(path)
  }

  private fun computeSampleSize(width: Int, height: Int, target: Int): Int {
    var sample = 1
    while (width / (sample * 2) >= target && height / (sample * 2) >= target) sample *= 2
    return sample
  }

  private fun applyExifOrientation(bitmap: Bitmap, orientation: Int): Bitmap {
    val matrix = Matrix()
    when (orientation) {
      ExifInterface.ORIENTATION_ROTATE_90 -> matrix.postRotate(90f)
      ExifInterface.ORIENTATION_ROTATE_180 -> matrix.postRotate(180f)
      ExifInterface.ORIENTATION_ROTATE_270 -> matrix.postRotate(270f)
      ExifInterface.ORIENTATION_FLIP_HORIZONTAL -> matrix.postScale(-1f, 1f)
      ExifInterface.ORIENTATION_FLIP_VERTICAL -> matrix.postScale(1f, -1f)
      else -> return bitmap
    }
    return Bitmap.createBitmap(bitmap, 0, 0, bitmap.width, bitmap.height, matrix, true)
  }

  private fun applyTransform(bitmap: Bitmap, transform: PhotoTransformState): Bitmap {
    if (transform.rotationDegrees == 0 && transform.straightenDegrees == 0f) return bitmap
    return Bitmap.createBitmap(bitmap, 0, 0, bitmap.width, bitmap.height, transform.toMatrix(), true)
  }

  private fun cropBitmap(bitmap: Bitmap, transform: PhotoTransformState): Bitmap {
    if (transform.cropLeft <= 0f && transform.cropTop <= 0f && transform.cropRight >= 1f && transform.cropBottom >= 1f) {
      return bitmap
    }
    val left = (transform.cropLeft * bitmap.width).roundToInt().coerceIn(0, bitmap.width - 1)
    val top = (transform.cropTop * bitmap.height).roundToInt().coerceIn(0, bitmap.height - 1)
    val right = (transform.cropRight * bitmap.width).roundToInt().coerceIn(left + 1, bitmap.width)
    val bottom = (transform.cropBottom * bitmap.height).roundToInt().coerceIn(top + 1, bitmap.height)
    return Bitmap.createBitmap(bitmap, left, top, right - left, bottom - top)
  }

  private fun resizeToFit(bitmap: Bitmap, maxWidth: Int?, maxHeight: Int?): Bitmap {
    if (maxWidth == null && maxHeight == null) return bitmap
    val widthLimit = maxWidth ?: bitmap.width
    val heightLimit = maxHeight ?: bitmap.height
    if (bitmap.width <= widthLimit && bitmap.height <= heightLimit) return bitmap
    val scale = minOf(widthLimit.toFloat() / bitmap.width, heightLimit.toFloat() / bitmap.height)
    val targetWidth = (bitmap.width * scale).roundToInt().coerceAtLeast(1)
    val targetHeight = (bitmap.height * scale).roundToInt().coerceAtLeast(1)
    return Bitmap.createScaledBitmap(bitmap, targetWidth, targetHeight, true)
  }

  private fun qualityFor(preset: String): Int = when (preset) {
    "low" -> 50
    "medium" -> 75
    "original" -> 100
    else -> 90
  }

  @Suppress("DEPRECATION")
  private fun legacyWebp() = Bitmap.CompressFormat.WEBP

  private fun encodingFor(format: String): Triple<Bitmap.CompressFormat, String, String> = when (format) {
    "png" -> Triple(Bitmap.CompressFormat.PNG, "image/png", "png")
    "webp" -> Triple(
      if (Build.VERSION.SDK_INT >= 30) Bitmap.CompressFormat.WEBP_LOSSY else legacyWebp(),
      "image/webp",
      "webp"
    )
    else -> Triple(Bitmap.CompressFormat.JPEG, "image/jpeg", "jpg")
  }

  private fun preserveMetadata(source: ExifInterface?, outputPath: String) {
    if (source == null) return
    try {
      val output = ExifInterface(outputPath)
      listOf(
        ExifInterface.TAG_DATETIME,
        ExifInterface.TAG_DATETIME_DIGITIZED,
        ExifInterface.TAG_MAKE,
        ExifInterface.TAG_MODEL,
        ExifInterface.TAG_GPS_LATITUDE,
        ExifInterface.TAG_GPS_LATITUDE_REF,
        ExifInterface.TAG_GPS_LONGITUDE,
        ExifInterface.TAG_GPS_LONGITUDE_REF
      ).forEach { tag -> source.getAttribute(tag)?.let { output.setAttribute(tag, it) } }
      // Pixels are already upright post-transform; the copied source orientation tag must not be reapplied.
      output.setAttribute(ExifInterface.TAG_ORIENTATION, ExifInterface.ORIENTATION_NORMAL.toString())
      output.saveAttributes()
    } catch (_: Exception) {
      // Best-effort only; a metadata write failure must not fail the export.
    }
  }
}
